### PR TITLE
Free variable annotation

### PR DESCRIPTION
**Pull request by ttuegel**
The `Valid` annotation is extended with a `freeVariables` field that tracks the free variables of a pattern. Substitution takes advantage of this annotation to avoid traversing and rewriting patterns unnecessarily.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by ttuegel** [comment]
See also: https://trello.com/c/4Qt2rJmA

File: src/main/haskell/kore/test/Test/Kore/Step/Pattern.hs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import           Test.Kore.Comparators ()
+import qualified Test.Kore.Step.MockSymbols as Mock
+import           Test.Tasty.HUnit.Extensions
+
+test_substitute :: [TestTree]
+test_substitute =
+    [ testCase "Variable: Target"
+        (assertEqualWithExplanation ""
+            (mkVar Mock.z)
+            (evalCounter $ substitute' (mkVar Mock.x))
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/test/Test/Kore/Step/Pattern.hs]: I'm not a fan of tests that need scrolling in order to be understood. I know that sometimes it's the best option, but here how about substituting `substitute'` with `substituteXWithZ` or inlining it?

File: src/main/haskell/kore/src/Kore/ASTVerifier/PatternVerifier.hs [hunk 2]
```diff
@@ -617,38 +696,48 @@ verifyVariable variable@Variable { variableName, variableSort } = do
...
     verified <-
         case domainValueChild of
             Domain.BuiltinPattern (StringLiteral_ _) -> sequence dv
             _ -> koreFail "Domain value argument must be a literal string."
-    return (Valid { patternSort } :< verified)
+    let freeVariables =
+            Foldable.foldl'
+                Set.union
+                Set.empty
+                (unifiedFreeVariables . extract <$> verified)
```
[the remark this answers is not in the document]
Comment (ttuegel) [reply src/main/haskell/kore/src/Kore/ASTVerifier/PatternVerifier.hs]: I will add the assertion; you are correct that `freeVariables` must be empty.

File: src/main/haskell/kore/src/Kore/Step/Pattern.hs [hunk 3]
```diff
@@ -11,19 +11,333 @@ module Kore.Step.Pattern
...
+toKorePattern =
+    Recursive.unfold toKorePatternWorker
+  where
+    toKorePatternWorker (Recursive.project -> ann :< pat) =
+        asUnified (Valid.mapVariables asUnified ann) :< asUnifiedPattern pat
+
+{- | Extract a 'StepPattern' from a 'KorePattern'.
+
+@patternKoreToStep@ does not lift the term, but rather fails with 'koreFail'
+any part of the pattern if of a different level.  For lifting functions see
```
[the remark this answers is not in the document]
Comment (ttuegel) [reply src/main/haskell/kore/src/Kore/Step/Pattern.hs]: That should read
```
... fails with 'koreFail' if any part of the pattern is on a different level.
```
Does that make more sense?

[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/Pattern.hs]: Yes, thank you.

**Pull request merged**